### DISCUSSION TITLE
How to share controller presets?

### DISCUSSION THREAD

**Discussion by tanders** [helgoboss/helgobox#373] [Q&A] [answered]
Another question: I have first/preliminary versions for controller presets for the "venerable" Behringer BCF2000 (using its factory preset 1) and the Nektar Panorama P6 (using its "Internal" mode, which turns it into a standard MIDI controller). Any interest in me sharing these? 

If so, where would I find the corresponding files on my machine? Also, are the respective projections included, or are these stored in separate files? How to submit these files? Would simply uploading them here be OK?

BTW: This projection feature is really great and is perhaps the feature that won me over for ReaLearn. :)

**Comment by helgoboss** [comment, marked as answer]
Would be awesome if you could share these. You can upload them here, yes. I'll have a look and add them to ReaPack with an author name of your choice. The projection layout is included in the JSON file.

Preset location (quoting from the user guide):

> All of your presets end up in the REAPER resource directory (REAPER → Actions → Show action list... → Show REAPER resource path in explorer/finder) at Data/helgoboss/realearn/presets. They are JSON files and very similar to what you get when you press Export to clipboard.

**Reply by tanders** [reply]
> Preset location (quoting from the user guide): 

Thanks a lot, I missed that. 

So, attached is my first attempt of a preset for the Behringer BCF2000 (https://www.soundonsound.com/reviews/behringer-bcf2000-bcr2000). This preset covers the Behringer factory `preset 1` of the controller, in which all the relevant encoders, buttons and faders send CC messages on channel 1. I mapped all the controllers that do send such messages, but left untouched those buttons that do not send any MIDI message in that preset either. The excluded buttons instead define 4 banks for the push encoders, allow to switch presets etc. I mapped all the banks, i.e. there are 32 encoders in the mapping, while there are only 8 hardware encoders on the device. 

 I think the actual mapping of this Behringer factory preset is fine and finished, but I am not yet 100% happy with the current projection layout. It just needs a lot of space currently. This device features push encoders, i.e. the rotary encoders additionally function as buttons, that send different CC than the encoder. In the projection, I represented all those push encoders by two GUI elements, an upper element representing an encoder, and a lower one the controller sent when the encoder is pressed like a button. I think ideally these two elements would be arranged in a concentric fashion, e.g., with the button being represented by an inner circle. However, I don't know how this would be best handled graphically in this interface such that the value feedback and the labels don't get in each other's ways.

Also, as mentioned I explicitly represented all 4 banks of the device directly. The advantage of that would be that one can see the mappings and current values of all the 32 encoders at the same time. On the other hand, this again takes rather much space: the four banks are represented in four rows (each controller as a pair, as discussed above). 

BTW: The lowest row of encoders in the projection is bank 1, bank 2 is above that and so forth. The idea here is that bank 1 is that way closer to the faders. So, in case one uses only 1 or 2 of the banks one can zoom in at the projection display so that only those rows are shown.

OK, below is the relevant link. Feedback on how this can be improved is most welcome. :) 

[bcf2000_preset_1.json.zip](https://github.com/helgoboss/realearn/files/6518063/bcf2000_preset_1.json.zip)

PS: In principle it is possible to turn all buttons on the device into controller elements and that way gain 10 more buttons for whatever mapping. I did not do this here, because when using a plain factory present there is then no need for others after downloading this Realearn controller preset to also download somewhere else some custom Behringer BCF2000 preset and to upload that to the device (which is a bit tricky). Perhaps more importantly even, if all buttons on the device are "captured" and used for custom mappings then it is not possible any more to switch Behringer BCF2000 presets on the device. So, doing something like that would only be interesting for people who want to fully "surrender" their device to ReaLearn. (The DrivenByMoss mappings happen to do exactly this, and I had figure out that I would first need to delete their Behringer BCF2000 preset before being able to use the device differently again.) I think the beauty of ReaLearn is that if you need more buttons you can simply get yourself another MIDI controller and add that!

**Reply by tanders** [reply]
BUG: I renamed the preset file without renaming the actual preset!

**Comment by tanders** [comment]
> With an author name of your choice

Just my real name is fine: Torsten Anders.

**Comment by tanders** [comment]
Here is a second controller preset, this time for the Nektar Panorama P6 (https://nektartech.com/panorama-p4-p6/). I mapped the so-called Internal mode of this device. This device can quickly switch between multiple DAW controller modes and one plain MIDI controller mode with dedicated buttons for that, the Internal mode turns the device into a plain MIDI controller. 

Again, I included mappings for all the controllers of the device (in Internal mode) that do send some MIDI CC message, but also pitch bend. In other words, I left unmapped the control elements sending MIDI notes such as the keys and pads, but also special keys like [Shift] and [F-Keys], the two toggle buttons, the [Pads] button, the menu and mode buttons and data encoder. 

This controller preset assumes some minor modifications of the Internal mode. All the rotary controllers have been switched from absolute to relative CC mode. Also, the single motor fader at the right was switched from CC 7 to CC 11.

Note that the shift key is not mapped directly. Instead its effect is used, so that the plain and shifted navigation buttons are contained in the mapping with their own respective CC messages. 

Also note that this device does not support MIDI feedback. The keyboard connects to Reaper via ReWire and additionally with a single MIDI input port, but no output port. 

I think the actual mappings of the control elements could again be considered finished, but I am again not quite happy with the arrangement of the controller element representations in the projection. These elements are currently arranged in a way that is very similar to how they are laid out on the actual device, but this results in a rather wide projection view. I am currently considering to quasi split this view in the middle and arrange the right half below the left half or the other way round so all elements are more compactly shown on something like a tablet. Could that confuse other users? Any better idea?

Anyway, below is the link. Again, comments for improvements are most welcome.

[nektar-panorama-p6.json.zip](https://github.com/helgoboss/realearn/files/6518422/nektar-panorama-p6.json.zip)

**Comment by helgoboss** [comment]
Thanks a lot, will try to check the presets soon! The additional info is very helpful.

**Comment by mangowannamango** [comment]
I'm a noob with this stuff, but it would be really cool (and attract more noobs, I think) if more presets could be made available, and put into a list for easy access. For instance, I use the Akai mpk mini mkII, and I think it's a relatively common MIDI controller. I just feel a little overwhelmed with how to use ReaLearn, and I wish I had a little more hand-holding. 

Great work overall though!

**Reply by helgoboss** [reply]
I suppose you are talking about "controller presets", not "main presets"? Controller presets just describe the controller, they don't do anything by itself because they don't contain any mappings to real targets (e.g. from a knob to track volume). Would be definitely great to have more of them built-in. 

Main presets would contain mappings to real targets, but as those are usually highly individual (especially for general-purpose controllers such as the Akai MPK), I'm not sure there's a point in having built-in ones (except of course presets for controlling Playtime, if applicable)?

In general, the addition of more built-in controller presets is mainly a question of time and money. I suspect that Playtime 2 will gradually improve the situation.

**Reply by mangowannamango** [reply]
At the time of writing, I didn't understand the difference between the controller compartment and the main compartment. I meant presets of any sort, but my brain was leaning more toward controller compartment stuff. It's more tedious for people who just want plug-and-play compatibility. 

I see your point about why it's less practical to share main compartment presets, but I disagree. By sharing them, it allows people to take a (mostly) functional instance and tweak it to their needs. Especially for controllers with similar layouts, wouldn't it just be a matter of re-learning the sources? (haha, I said Re-Learn). Basically, on the assumption that you don't have enough time to post "use-case"-style tutorials (format: "How to {do a specific type of workflow, such as switch VST sounds on one track with one button push}"),  sharing main compartment presets would allow us to at least be able to (maybe) figure it out ourselves.

Also, see [my much more thorough post](https://github.com/helgoboss/helgobox/discussions/1078) begging for better user support for ReaLearn.

No offense, just curious. What does Playtime have to do with this?

Thanks!

**Reply by mangowannamango** [reply]
Also, I thought I'd contribute my Akai mpk mini mk2 controller compartment preset. Putting this message here for now. When I figure out how to do it, I'll add the file here.

**Reply by helgoboss** [reply]
> I see your point about why it's less practical to share main compartment presets, but I disagree. By sharing them, it allows people to take a (mostly) functional instance and tweak it to their needs. Especially for controllers with similar layouts, wouldn't it just be a matter of re-learning the sources? (haha, I said Re-Learn). Basically, on the assumption that you don't have enough time to post "use-case"-style tutorials (format: "How to {do a specific type of workflow, such as switch VST sounds on one track with one button push}"), sharing main compartment presets would allow us to at least be able to (maybe) figure it out ourselves.

Okay, but what would those main presets do? In case of an MPK, for instance? Is there any good default what those knobs and pads should control?

> No offense, just curious. What does Playtime have to do with this?

Playtime has been the driving force of Helgobox development since one year and still will be until about end of the year. The controller integration for Playtime is powered by ReaLearn. There's a growing collection of official presets for popular grid controllers (both universal controller presets and Playtime-tailored main presets). And after covering the grid controllers, I probably will also gradually add official presets for keyboards that have a sort of grid or at least pads.
